reddit_api_async/connection.ml

```diff
@@ -593,13 +635,22 @@ module For_testing = struct
 
     let with_t filename ~credentials ~f =
       let placeholders = Placeholders.create () in
-      let ({ client_id; client_secret; password; username } : Credentials.t) =
-        credentials
-      in
-      Placeholders.add placeholders ~secret:client_id ~placeholder:"client_id";
-      Placeholders.add placeholders ~secret:client_secret ~placeholder:"client_secret";
-      Placeholders.add placeholders ~secret:password ~placeholder:"password";
-      Placeholders.add placeholders ~secret:username ~placeholder:"username";
+      (match (credentials : Credentials.t) with
+      | Password { client_id; client_secret; username; password } ->
+        Placeholders.add placeholders ~secret:client_id ~placeholder:"client_id";
+        Placeholders.add placeholders ~secret:client_secret ~placeholder:"client_secret";
+        Placeholders.add placeholders ~secret:password ~placeholder:"password";
+        Placeholders.add placeholders ~secret:username ~placeholder:"username"
+      | Userless_confidential { client_id; client_secret } ->
+        Placeholders.add placeholders ~secret:client_id ~placeholder:"client_id";
+        Placeholders.add placeholders ~secret:client_secret ~placeholder:"client_secret"
+      | Userless_public { client_id; device_id } ->
+        Placeholders.add placeholders ~secret:client_id ~placeholder:"client_id";
+        Placeholders.add
+          placeholders
+          ~secret:
+            (Credentials.Userless_public.device_id_or_default { client_id; device_id })
```

Comment:
I thought `as` pattern might work but it didn't, so made `Credentials.Userless_public.t` from the destructured fields. Please let me know if there's any better way.

Reply:
What didn't work, exactly? Were you by any chance trying without parentheses? `| Userless_public ({ client_id; device_id = _ } as v) ->` will work, but `| Userless_public { client_id; device_id = _ } as v ->` is problematic because `v` would be the whole `Credentials.t` and rather than the `Credentials.Userless_public.t`.

Reply:
> What didn't work, exactly? Were you by any chance trying without parentheses? 

Exactly. Your code worked perfectly. I haven't thought that parens can be put on that position.